File: src/lsp/tree.go
// The tree handed in. Every check reads the disk through this, so a test drives
// the same code over a folder it writes itself, and the editor drives it over
// the buffer a person is typing into.
// [[spec/design_output/tree#the-tree-handed-in]]
package main

import (
	"encoding/json"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"sync"
)

type Box struct {
	User  string
	Home  string
	Name  string
	Email string
}

type Tree struct {
	Root  string
	Words int
	Node  string
	Box   Box

	guard   sync.Mutex
	overlay map[string]string
	held    []string
}

func treeAt(root string) *Tree {
	return &Tree{Root: root, overlay: map[string]string{}}
}

// Holds parks the text of a buffer the editor owns, so a check reads what a
// person types before the disk carries it.
func (one *Tree) Holds(path, text string) {
	one.guard.Lock()
	defer one.guard.Unlock()
	one.overlay[slashed(path)] = text
	one.held = nil
}

func (one *Tree) Drops(path string) {
	one.guard.Lock()
	defer one.guard.Unlock()
	delete(one.overlay, slashed(path))
	one.held = nil
}

func (one *Tree) Read(path string) string {
	said := slashed(path)
	one.guard.Lock()
	text, open := one.overlay[said]
	one.guard.Unlock()
	if open {
		return text
	}
	read, err := os.ReadFile(filepath.Join(one.Root, filepath.FromSlash(said)))
	if err != nil {
		return ""
	}
	return string(read)
}

func (one *Tree) Exists(path string) bool {
	_, err := os.Stat(filepath.Join(one.Root, filepath.FromSlash(path)))
	return err == nil
}

func (one *Tree) Names(folder, end string) []string {
	found, err := os.ReadDir(filepath.Join(one.Root, filepath.FromSlash(folder)))
	if err != nil {
		return nil
	}
	out := []string{}
	for _, entry := range found {
		if !entry.IsDir() && strings.HasSuffix(entry.Name(), end) {
			out = append(out, entry.Name())
		}
	}
	sort.Strings(out)
	return out
}

// [[spec/design_output/tree#the-tree-handed-in]]
func (one *Tree) Paths() []string {
	one.guard.Lock()
	defer one.guard.Unlock()
	if one.held != nil {
		return one.held
	}

	out := gitHolds(one.Root)
	if out == nil {
		out = diskHolds(one.Root)
	}
	one.held = out
	return out
}

// [[spec/design_output/tree#the-tree-handed-in]]
func gitHolds(root string) []string {
	said := exec.Command("git", "ls-files")
	said.Dir = root
	read, err := said.Output()
	if err != nil {
		return nil
	}
	out := []string{}
	for _, line := range splitLines(string(read)) {
		path := strings.TrimSpace(line)
		if path != "" && !isDraft(path) {
			out = append(out, path)
		}
	}
	return out
}

// A clone with no git walks the disk, so every rule still answers.
// [[spec/design_output/tree#the-tree-handed-in]]
func diskHolds(root string) []string {
	out := []string{}
	filepath.WalkDir(root, func(where string, entry os.DirEntry, err error) error {
		if err != nil {
			return nil
		}
		name := entry.Name()
		if entry.IsDir() {
			if where != root && (name == ".git" || name == ".se" || name == "node_modules") {
				return filepath.SkipDir
			}
			return nil
		}
		said, err := filepath.Rel(root, where)
		if err != nil || isDraft(said) {
			return nil
		}
		out = append(out, slashed(said))
		return nil
	})
	sort.Strings(out)
	return out
}

// Forgets drops the walk, so the next sweep reads what git holds now.
func (one *Tree) Forgets() {
	one.guard.Lock()
	defer one.guard.Unlock()
	one.held = nil
}

const (
	Install  = "src/scripts/install.sh"
	ValeIni  = ".vale.ini"
	Settings = ".vscode/settings.json"
	Offered  = ".vscode/extensions.json"
	ToolsAt  = ".se/tools.json"
	Bin      = ".se/bin"
)

// [[spec/design_output/editor#what-the-editor-runs]]
var Extensions = []string{"chrischinchilla.vale-vscode", "biomejs.biome"}

// [[spec/design_output/tools#what-the-survey-writes]]
var Wanted = []string{"node", "vale", "biome", "vale-ls", "se-lsp", "go", "git", "claude", "sh", "python"}

// [[spec/design_output/editor#what-the-tracked-settings-say]]
func settingsNameBinaries(tree *Tree) []Finding {
	rule := "SettingsNameBinaries"
	text := tree.Read(Settings)
	said := parsedJSON(text)
	if said == nil {
		return []Finding{unread(rule, Settings)}
	}

	out := []Finding{}
	for _, one := range namesTheBinaries(said) {
		if one.holds {
			continue
		}
		out = append(out, fault(rule, Settings, lineOf(text, one.key), one.key+" names something else. "+one.why))
	}

	installs := installedTools(tree.Read(Install))
	for _, name := range []string{"vale", "biome"} {
		if has(installs, name) {
			continue
		}
		out = append(out, fault(rule, Install, 1,
			Settings+" runs "+Bin+"/"+name+", and this script installs no "+name+"."))
	}
	return out
}

type binaryCheck struct {
	key   string
	why   string
	holds bool
}

func namesTheBinaries(said map[string]any) []binaryCheck {
	biome, holdsBiome := said["biome.lsp.bin"]
	paths := []string{}
	switch one := biome.(type) {
	case string:
		paths = append(paths, one)
	case map[string]any:
		for _, key := range keysOf(one) {
			paths = append(paths, asText(one[key]))
		}
	}
	named := len(paths) > 0
	for _, one := range paths {
		if !strings.HasPrefix(one, Bin+"/biome") {
			named = false
		}
	}
	_ = holdsBiome

	return []binaryCheck{
		{"vale.valeCLI.path", "The editor runs " + Bin + "/vale, which " + Install + " writes.",
			asText(said["vale.valeCLI.path"]) == Bin+"/vale"},
		{"vale.valeCLI.config", "The editor reads " + ValeIni + " at the root.",
			asText(said["vale.valeCLI.config"]) == ValeIni},
		{"vale.valeCLI.installVale", Install + " pins Vale, so the extension installs none of its own.",
			said["vale.valeCLI.installVale"] == false},
		{"biome.lsp.bin", "The editor runs " + Bin + "/biome, which " + Install + " writes.", named},
		{"biome.configurationPath", "Biome reads spec/config/biome.json, which this tree tracks.",
			asText(said["biome.configurationPath"]) == "spec/config/biome.json"},
	}
}

// [[spec/design_output/editor#what-the-editor-runs]]
func editorDrawsWriteRules(tree *Tree) []Finding {
	rule := "EditorDrawsWriteRules"
	text := tree.Read(Settings)
	said := parsedJSON(text)
	if said == nil {
		return []Finding{unread(rule, Settings)}
	}

	out := []Finding{}
	where := asText(said["vale.valeCLI.config"])
	ini := ""
	if where != "" {
		ini = tree.Read(where)
	}
	if ini == "" {
		names := where
		if names == "" {
			names = "nothing"
		}
		return append(out, fault(rule, Settings, lineOf(text, "vale.valeCLI.config"),
			"vale.valeCLI.config names "+names+", and the write door reads "+ValeIni+"."))
	}

	level := valeLevel(ini)
	if asText(said["vale.valeCLI.minAlertLevel"]) != "inherited" {
		drawn := level
		if drawn == "" {
			drawn = "its own level"
		}
		out = append(out, fault(rule, Settings, lineOf(text, "vale.valeCLI.minAlertLevel"),
			where+" draws at "+drawn+". Set vale.valeCLI.minAlertLevel to inherited."))
	}
	if !spellingStyle(ini) && said["vale.enableSpellcheck"] != false {
		out = append(out, fault(rule, Settings, lineOf(text, "vale.enableSpellcheck"),
			where+" names no spelling style. Set vale.enableSpellcheck to false."))
	}
	if said["vale.valeCLI.lintOnChange"] != true {
		out = append(out, fault(rule, Settings, lineOf(text, "vale.valeCLI.lintOnChange"),
			"Set vale.valeCLI.lintOnChange to true, so a rule draws while a person types."))
	}
	return out
}

// [[spec/design_output/editor#what-the-tracked-settings-say]]
func biomeOnWindows(tree *Tree) []Finding {
	rule := "BiomeOnWindows"
	text := tree.Read(Settings)
	said := parsedJSON(text)
	if said == nil {
		return []Finding{unread(rule, Settings)}
	}

	held, named := said["biome.lsp.bin"].(map[string]any)
	if !named {
		return []Finding{fault(rule, Settings, lineOf(text, "biome.lsp.bin"),
			"biome.lsp.bin names one path per platform, under "+Bin+".")}
	}

	out := []Finding{}
	for _, platform := range keysOf(held) {
		path := asText(held[platform])
		wants := Bin + "/biome"
		if strings.HasPrefix(platform, "win32") {
			wants = Bin + "/biome.exe"
		}
		if path == wants {
			continue
		}
		out = append(out, fault(rule, Settings, lineOf(text, platform),
			platform+" runs "+path+", and this tree installs "+wants+" there."))
	}
	return out
}

// [[spec/design_output/editor#what-the-tracked-settings-say]]
func extensionsOnOffer(tree *Tree) []Finding {
	rule := "ExtensionsOnOffer"
	text := tree.Read(Offered)
	said := parsedJSON(text)
	if said == nil {
		return []Finding{unread(rule, Offered)}
	}

	out := []Finding{}
	offered := []string{}
	for _, one := range listOf(said["recommendations"]) {
		offered = append(offered, asText(one))
	}
	where := lineOf(text, "recommendations")
	for _, one := range Extensions {
		if !has(offered, one) {
			out = append(out, fault(rule, Offered, where, "A clone opens without "+one+" on offer."))
		}
	}
	for _, one := range offered {
		if !has(Extensions, one) {
			out = append(out, fault(rule, Offered, where, one+" holds no rule this tree reads."))
		}
	}

	settings := tree.Read(Settings)
	held := parsedJSON(settings)
	for _, key := range keysOf(held) {
		under, nested := held[key].(map[string]any)
		if !nested {
			continue
		}
		formatter := asText(under["editor.defaultFormatter"])
		if formatter == "" || has(offered, formatter) {
			continue
		}
		out = append(out, fault(rule, Settings, lineOf(settings, formatter),
			key+" formats through "+formatter+", and "+Offered+" offers no such extension."))
	}
	return out
}

// [[spec/design_output/log#nothing-here-deletes-a-log]]
func noLogDeleted(tree *Tree) []Finding {
	rule := "NoLogDeleted"
	out := []Finding{}
	for _, path := range tree.Paths() {
		if !sourceFile(path) {
			continue
		}
		for i, line := range splitLines(tree.Read(path)) {
			if !deletesAt.MatchString(line) || !loggedAt.MatchString(line) {
				continue
			}
			out = append(out, fault(rule, path, i+1,
				"This line reaches a log file. The log is the record of what every door does, so nothing here takes one away."))
		}
	}
	return out
}

// [[spec/design_output/level0#a-name-holds-five-words]]
func nameHoldsTheWords(tree *Tree) []Finding {
	rule := "NameHoldsTheWords"
	out := []Finding{}
	if tree.Words == 0 {
		return out
	}
	for _, path := range tree.Paths() {
		part := overLong(path, tree.Words)
		if part == "" {
			continue
		}
		out = append(out, fault(rule, path, 1,
			part+" holds more than "+strconv.Itoa(tree.Words)+" words. Rename it shorter."))
	}
	return out
}

// [[spec/design_output/private#the-box-names-the-owner]]
func nothingPrivateTravels(tree *Tree) []Finding {
	rule := "NothingPrivateTravels"
	home := strings.TrimRight(slashed(tree.Box.Home), "/")
	if !homeNames(home) {
		home = ""
	}

	type wanted struct{ what, said string }
	held := []wanted{}
	for _, one := range []wanted{
		{"the user this box runs as", tree.Box.User},
		{"the home folder on this box", home},
		{"the git name on this box", tree.Box.Name},
		{"the git address on this box", tree.Box.Email},
	} {
		if namesAPerson(one.said) {
			held = append(held, one)
		}
	}
	if len(held) == 0 {
		return nil
	}

	out := []Finding{}
	for _, path := range tree.Paths() {
		if !textFile(path) {
			continue
		}
		for i, line := range splitLines(tree.Read(path)) {
			for _, one := range held {
				if !carriesTheName(line, one.said) {
					continue
				}
				out = append(out, fault(rule, path, i+1,
					"This line carries "+one.what+", and git carries this file everywhere. "+
						"Say what the thing is, in words a reader outside this box acts on."))
			}
		}
	}
	return out
}

// [[spec/design_output/tools#what-the-survey-names]]
func surveyNamesInstalls(tree *Tree) []Finding {
	rule := "SurveyNamesInstalls"
	text := tree.Read(Install)
	installs := installedTools(text)
	if len(installs) == 0 {
		return []Finding{fault(rule, Install, 1, "This script names no tool the survey reads back.")}
	}

	out := []Finding{}
	for _, name := range installs {
		if has(Wanted, name) {
			continue
		}
		out = append(out, fault(rule, Install, lineOf(text, name+")"),
			"The survey names no "+name+", so every caller guesses its path. Add it to WANTED."))
	}
	return out
}

// [[spec/design_output/tools#what-the-survey-writes]]
func surveyFindsNode(tree *Tree) []Finding {
	rule := "SurveyFindsNode"
	text := tree.Read(ToolsAt)
	if text == "" {
		return []Finding{fault(rule, Install, 1,
			ToolsAt+" stands nowhere, so every caller guesses a path. Run ./RUNME.sh tools.")}
	}
	// The sweep names the node that runs it, and no node runs this server.
	if tree.Node == "" {
		return nil
	}

	said := ""
	if node, held := parsedJSON(text)["node"].(map[string]any); held {
		said = asText(node["version"])
	}
	if said == tree.Node {
		return nil
	}
	names := said
	if names == "" {
		names = "nothing"
	}
	return []Finding{fault(rule, ToolsAt, lineOf(text, "version"),
		"The survey names node "+names+", and node "+tree.Node+" runs this sweep. Run ./RUNME.sh tools.")}
}

// [[spec/design_output/tree#what-a-rule-answers]]
var Rules = []func(*Tree) []Finding{
	settingsNameBinaries,
	editorDrawsWriteRules,
	biomeOnWindows,
	extensionsOnOffer,
	noLogDeleted,
	nameHoldsTheWords,
	nothingPrivateTravels,
	surveyNamesInstalls,
	surveyFindsNode,
}

// [[spec/design_output/tree#what-a-rule-answers]]
func treeFaults(tree *Tree) []Finding {
	out := []Finding{}
	for _, rule := range Rules {
		out = append(out, rule(tree)...)
	}
	return out
}

func unread(rule, where string) Finding {
	return fault(rule, where, 1,
		"This file reads as no JSON, so every rule over it stands unchecked.")
}

func parsedJSON(text string) map[string]any {
	if strings.TrimSpace(text) == "" {
		return nil
	}
	var said map[string]any
	if err := json.Unmarshal([]byte(text), &said); err != nil {
		return nil
	}
	return said
}

func keysOf(said map[string]any) []string {
	out := make([]string, 0, len(said))
	for key := range said {
		out = append(out, key)
	}
	sort.Strings(out)
	return out
}

func listOf(said any) []any {
	one, held := said.([]any)
	if !held {
		return nil
	}
	return one
}

func asText(said any) string {
	one, held := said.(string)
	if !held {
		return ""
	}
	return one
}

func itoa(said int) string { return strconv.Itoa(said) }

func has(said []string, one string) bool {
	for _, each := range said {
		if each == one {
			return true
		}
	}
	return false
}

func lineOf(text, needle string) int {
	if needle == "" {
		return 1
	}
	for i, line := range splitLines(text) {
		if strings.Contains(line, needle) {
			return i + 1
		}
	}
	return 1
}

File: .claude/skills/level0/lib/tree.js
// Every rule weighing one file against another. Vale hands a rule one buffer,
// so a rule reading two files lives here, and the command line runs it over the
// whole tree into the problems panel.
// [[spec/design_output/tree#the-rules-over-two-files]]

import { overLong } from "./names.js";
import { isDraft } from "./paths.js";
import { carriesTheName, namesAPerson } from "./private.js";
import {
  EDITOR_EXTENSIONS,
  EDITOR_SETTINGS,
  EXTENSIONS,
  namesTheBinaries,
} from "./servers.js";
import { decide, pool, RULES as STOP } from "./stop.js";
import { BIN, installedTools, TOOLS, WANTED } from "./tools.js";

export const INSTALL = "src/scripts/install.sh";
export const VALE_INI = ".vale.ini";

const STOP_LIB = ".claude/skills/level0/lib/stop.js";
const SOURCE = /^(?:src|\.claude)\/.*\.js$/;
const TEXT = /\.(?:md|markdown|txt|ya?ml|json|js|ts|tsx|go|sh|ps1|ini|mod)$/i;
const DELETES = /\bremove\(|\bunlink|\brm\b|\bprune\b/;
const LOGGED = /log/i;

const LATER = {
  name: "level1.yml",
  text:
    "- id: a-later-level\n  side: stop\n  priority: 20\n  decides: claimed\n" +
    "  asks: Later?\n  says: A later level says so.\n",
};

const BINARIES = {
  vale: ["vale.valeCLI.path", `The editor runs ${BIN}/vale, which ${INSTALL} writes.`],
  valeConfig: ["vale.valeCLI.config", `The editor reads ${VALE_INI} at the root.`],
  managesVale: [
    "vale.valeCLI.installVale",
    `${INSTALL} pins Vale, so the extension installs none of its own.`,
  ],
  biome: ["biome.lsp.bin", `The editor runs ${BIN}/biome, which ${INSTALL} writes.`],
  biomeConfig: [
    "biome.configurationPath",
    "Biome reads spec/config/biome.json, which this tree tracks.",
  ],
};

// [[spec/design_output/tree#the-tree-handed-in]]
export function treeOf(it) {
  const at = (path) => `${it.root}/${path}`;
  let held = null;
  return {
    words: it.words ?? 0,
    node: it.node ?? "",
    box: it.box ?? {},
    read(path) {
      try {
        return it.disk.read(at(path));
      } catch {
        return "";
      }
    },
    exists: (path) => it.disk.exists(at(path)),
    names(folder, end) {
      try {
        return it.disk
          .list(at(folder))
          .filter((one) => one.kind === "file" && one.name.endsWith(end))
          .map((one) => one.name);
      } catch {
        return [];
      }
    },
    paths() {
      if (held) return held;
      held = it.git
        .run(["ls-files"], true)
        .out.split(/\r?\n/)
        .map((one) => one.trim())
        .filter(Boolean)
        .filter((one) => !isDraft(one));
      return held;
    },
  };
}

// [[spec/design_output/editor#what-the-tracked-settings-say]]
export function settingsNameBinaries(tree) {
  const rule = "SettingsNameBinaries";
  const text = tree.read(EDITOR_SETTINGS);
  const said = parsed(text);
  if (!said) return [unread(rule, EDITOR_SETTINGS)];

  const out = [];
  for (const [flag, holds] of Object.entries(namesTheBinaries(said))) {
    if (holds) continue;
    const [key, why] = BINARIES[flag];
    out.push(
      fault(
        rule,
        EDITOR_SETTINGS,
        `${key} names something else. ${why}`,
        lineOf(text, key),
      ),
    );
  }

  const installs = installedTools(tree.read(INSTALL));
  for (const name of ["vale", "biome"]) {
    if (installs.includes(name)) continue;
    out.push(
      fault(
        rule,
        INSTALL,
        `${EDITOR_SETTINGS} runs ${BIN}/${name}, and this script installs no ${name}.`,
      ),
    );
  }
  return out;
}

// [[spec/design_output/editor#what-the-editor-runs]]
export function editorDrawsWriteRules(tree) {
  const rule = "EditorDrawsWriteRules";
  const text = tree.read(EDITOR_SETTINGS);
  const said = parsed(text);
  if (!said) return [unread(rule, EDITOR_SETTINGS)];

  const out = [];
  const where = said["vale.valeCLI.config"] ?? "";
  const ini = where ? tree.read(where) : "";
  if (!ini) {
    out.push(
      fault(
        rule,
        EDITOR_SETTINGS,
        `vale.valeCLI.config names ${where || "nothing"}, and the write door reads ${VALE_INI}.`,
        lineOf(text, "vale.valeCLI.config"),
      ),
    );
    return out;
  }

  const level = /^\s*MinAlertLevel\s*=\s*(\S+)/m.exec(ini)?.[1] ?? "";
  if (said["vale.valeCLI.minAlertLevel"] !== "inherited") {
    out.push(
      fault(
        rule,
        EDITOR_SETTINGS,
        `${where} draws at ${level || "its own level"}. Set vale.valeCLI.minAlertLevel to inherited.`,
        lineOf(text, "vale.valeCLI.minAlertLevel"),
      ),
    );
  }
  if (!/BasedOnStyles.*Spelling/.test(ini) && said["vale.enableSpellcheck"] !== false) {
    out.push(
      fault(
        rule,
        EDITOR_SETTINGS,
        `${where} names no spelling style. Set vale.enableSpellcheck to false.`,
        lineOf(text, "vale.enableSpellcheck"),
      ),
    );
  }
  if (said["vale.valeCLI.lintOnChange"] !== true) {
    out.push(
      fault(
        rule,
        EDITOR_SETTINGS,
        "Set vale.valeCLI.lintOnChange to true, so a rule draws while a person types.",
        lineOf(text, "vale.valeCLI.lintOnChange"),
      ),
    );
  }
  return out;
}

// [[spec/design_output/editor#what-the-tracked-settings-say]]
export function biomeOnWindows(tree) {
  const rule = "BiomeOnWindows";
  const text = tree.read(EDITOR_SETTINGS);
  const said = parsed(text);
  if (!said) return [unread(rule, EDITOR_SETTINGS)];

  const map = said["biome.lsp.bin"];
  if (!map || typeof map !== "object") {
    return [
      fault(
        rule,
        EDITOR_SETTINGS,
        `biome.lsp.bin names one path per platform, under ${BIN}.`,
        lineOf(text, "biome.lsp.bin"),
      ),
    ];
  }

  const out = [];
  for (const [platform, path] of Object.entries(map)) {
    const wants = platform.startsWith("win32") ? `${BIN}/biome.exe` : `${BIN}/biome`;
    if (path === wants) continue;
    out.push(
      fault(
        rule,
        EDITOR_SETTINGS,
        `${platform} runs ${path}, and this tree installs ${wants} there.`,
        lineOf(text, platform),
      ),
    );
  }
  return out;
}

// [[spec/design_output/editor#what-the-tracked-settings-say]]
export function extensionsOnOffer(tree) {
  const rule = "ExtensionsOnOffer";
  const text = tree.read(EDITOR_EXTENSIONS);
  const said = parsed(text);
  if (!said) return [unread(rule, EDITOR_EXTENSIONS)];

  const out = [];
  const offered = Array.isArray(said.recommendations) ? said.recommendations : [];
  const at = lineOf(text, "recommendations");
  for (const one of EXTENSIONS) {
    if (offered.includes(one)) continue;
    out.push(
      fault(rule, EDITOR_EXTENSIONS, `A clone opens without ${one} on offer.`, at),
    );
  }
  for (const one of offered) {
    if (EXTENSIONS.includes(one)) continue;
    out.push(
      fault(rule, EDITOR_EXTENSIONS, `${one} holds no rule this tree reads.`, at),
    );
  }

  const settings = tree.read(EDITOR_SETTINGS);
  for (const [key, value] of Object.entries(parsed(settings) ?? {})) {
    const formatter = value?.["editor.defaultFormatter"];
    if (!formatter || offered.includes(formatter)) continue;
    out.push(
      fault(
        rule,
        EDITOR_SETTINGS,
        `${key} formats through ${formatter}, and ${EDITOR_EXTENSIONS} offers no such extension.`,
        lineOf(settings, formatter),
      ),
    );
  }
  return out;
}

// [[spec/design_output/stop#where-the-rules-live]]
export function stopFolderIsData(tree) {
  const rule = "StopFolderIsData";
  const mine = tree
    .names(STOP, ".yml")
    .map((name) => ({ name, text: tree.read(`${STOP}/${name}`) }));
  const one = pool(mine);

  const out = [];
  for (const name of one.broken) {
    out.push(
      fault(
        rule,
        `${STOP}/${name}`,
        "This file carries no whole rule, so the pool drops it.",
      ),
    );
  }

  const both = pool([...mine, LATER]);
  if (both.rules.length !== one.rules.length + 1) {
    out.push(
      fault(
        rule,
        STOP_LIB,
        "A second file in the folder adds no rule, so the pool reads code.",
      ),
    );
    return out;
  }
  const said = decide(both.rules, { claimed: "a-later-level", ran: () => false });
  if (!said.ends) {
    out.push(
      fault(
        rule,
        STOP_LIB,
        "A rule out of a second file never fires, so the vote reads code.",
      ),
    );
  }
  return out;
}

// [[spec/design_output/log#nothing-here-deletes-a-log]]
export function noLogDeleted(tree) {
  const rule = "NoLogDeleted";
  const out = [];
  for (const path of tree.paths().filter((one) => SOURCE.test(one))) {
    const lines = tree.read(path).split(/\r?\n/);
    for (let i = 0; i < lines.length; i++) {
      if (!DELETES.test(lines[i]) || !LOGGED.test(lines[i])) continue;
      out.push(
        fault(
          rule,
          path,
          "This line reaches a log file. The log is the record of what every door does, so nothing here takes one away.",
          i + 1,
        ),
      );
    }
  }
  return out;
}

// [[spec/design_output/level0#a-name-holds-five-words]]
export function nameHoldsTheWords(tree) {
  const rule = "NameHoldsTheWords";
  const out = [];
  if (!tree.words) return out;
  for (const path of tree.paths()) {
    const part = overLong(path, tree.words);
    if (!part) continue;
    out.push(
      fault(
        rule,
        path,
        `${part} holds more than ${tree.words} words. Rename it shorter.`,
      ),
    );
  }
  return out;
}

// [[spec/design_output/private#the-box-names-the-owner]]
export function nothingPrivateTravels(tree) {
  const rule = "NothingPrivateTravels";
  const box = tree.box ?? {};
  const home = String(box.home ?? "").replace(/[/\\]+$/, "");

  const wanted = [
    ["the user this box runs as", box.user],
    ["the home folder on this box", homeNames(home) ? home : ""],
    ["the git name on this box", box.name],
    ["the git address on this box", box.email],
  ].filter(([, said]) => namesAPerson(said));
  if (!wanted.length) return [];

  const out = [];
  for (const path of tree.paths().filter((one) => TEXT.test(one))) {
    const lines = tree.read(path).split(/\r?\n/);
    for (let i = 0; i < lines.length; i++) {
      for (const [what, said] of wanted) {
        if (!carriesTheName(lines[i], said)) continue;
        out.push(
          fault(
            rule,
            path,
            `This line carries ${what}, and git carries this file everywhere. Say what the thing is, in words a reader outside this box acts on.`,
            i + 1,
          ),
        );
      }
    }
  }
  return out;
}

function homeNames(home) {
  const who = String(home).split(/[/\\]+/).filter(Boolean).pop() ?? "";
  return namesAPerson(who);
}

// [[spec/design_output/tools#what-the-survey-names]]
export function surveyNamesInstalls(tree) {
  const rule = "SurveyNamesInstalls";
  const text = tree.read(INSTALL);
  const installs = installedTools(text);
  if (!installs.length) {
    return [fault(rule, INSTALL, "This script names no tool the survey reads back.")];
  }

  const out = [];
  const wanted = WANTED.map((one) => one.name);
  for (const name of installs) {
    if (wanted.includes(name)) continue;
    out.push(
      fault(
        rule,
        INSTALL,
        `The survey names no ${name}, so every caller guesses its path. Add it to WANTED.`,
        lineOf(text, `${name})`),
      ),
    );
  }
  return out;
}

// [[spec/design_output/tools#what-the-survey-writes]]
export function surveyFindsNode(tree) {
  const rule = "SurveyFindsNode";
  const text = tree.read(TOOLS);
  if (!text) {
    return [
      fault(
        rule,
        INSTALL,
        `${TOOLS} stands nowhere, so every caller guesses a path. Run ./RUNME.sh tools.`,
      ),
    ];
  }

  const said = parsed(text)?.node?.version ?? "";
  if (said === tree.node) return [];
  return [
    fault(
      rule,
      TOOLS,
      `The survey names node ${said || "nothing"}, and node ${tree.node} runs this sweep. Run ./RUNME.sh tools.`,
      lineOf(text, "version"),
    ),
  ];
}

export const RULES = [
  settingsNameBinaries,
  editorDrawsWriteRules,
  biomeOnWindows,
  extensionsOnOffer,
  stopFolderIsData,
  noLogDeleted,
  nameHoldsTheWords,
  nothingPrivateTravels,
  surveyNamesInstalls,
  surveyFindsNode,
];

// [[spec/design_output/tree#what-a-rule-answers]]
export function treeFaults(tree) {
  const out = [];
  for (const rule of RULES) out.push(...rule(tree));
  return out;
}

function fault(rule, file, message, line = 1) {
  return { file, rule, line, column: 1, message, severity: "error" };
}

function unread(rule, where) {
  return fault(
    rule,
    where,
    "This file reads as no JSON, so every rule over it stands unchecked.",
  );
}

function parsed(text) {
  try {
    return JSON.parse(text);
  } catch {
    return null;
  }
}

function lineOf(text, needle) {
  const lines = String(text).split(/\r?\n/);
  for (let i = 0; i < lines.length; i++) {
    if (lines[i].includes(needle)) return i + 1;
  }
  return 1;
}
